### PR TITLE
Implemented StaticFileConverter to migrate static files

### PR DESCRIPTION
**Pull request by KallanBao**
## Description
* Implemented StaticFileConverter class to migrate static files to wwwroot directory
* Code refactoring, FileInformation objects now FileConverters, created new FileInformation objects

## Supplemental testing
Basic unit testing to make sure file byte array representation is generated and relative path is changed accordingly

---
*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: src/CTA.WebForms2Blazor/FileConverters/StaticFileConverter.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+            var workingDirectory = Environment.CurrentDirectory;
+            var testProjectPath = Directory.GetParent(workingDirectory).Parent.Parent.FullName;
+            var path = Path.Combine(testProjectPath, RelativePath);
+            return File.ReadAllBytes(path);
+        }
+
+        public override async Task<IEnumerable<FileInformation>> MigrateFileAsync()
+        {
+            var newPath = Path.Combine("wwwroot", RelativePath);
+            FileInformation fi = new StaticFileInformation(newPath, GetFileBytes());
```
Comment (comsemer) [inline src/CTA.WebForms2Blazor/FileConverters/StaticFileConverter.cs:29]: This can just create an instance of FileInformation once it is no longer abstract, there's no need to use a subclass if it offers no extra functionality

File: tst/CTA.WebForms2Blazor.Tests/FileConverterTests.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using CTA.WebForms2Blazor.FileConverters;
+
+namespace CTA.WebForms2Blazor.Tests
+{
+    [TestFixture]
+    class FileConverterTests
+    {
+        private const string TestFilesDirectoryPath = "TestingArea\\TestFiles";
```
Comment (jonlouie) [inline tst/CTA.WebForms2Blazor.Tests/FileConverterTests.cs:18]: Same as above, use `Path.Combine()`

File: tst/CTA.WebForms2Blazor.Tests/FileConverterTests.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+            _testStaticFilePath = Path.Combine(TestFilesDirectoryPath, "SampleStaticFile.png");
+            _testViewFilePath = Path.Combine(TestFilesDirectoryPath, "SampleViewFile.aspx");
+            _testProjectFilePath = Path.Combine(TestFilesDirectoryPath, "SampleProjectFile.csproj");
+        }
+
+        [SetUp]
+        public void Setup()
+        {
+
+        }
```
Comment (jonlouie) [inline tst/CTA.WebForms2Blazor.Tests/FileConverterTests.cs:42]: Can remove if not being used. If being saved for future use, leave a quick comment inside

**Review by comsemer — CHANGES_REQUESTED** [review]
Mostly good, I just found some general things that I think should be altered regarding the new File Information and File Converter classes

File: tst/CTA.WebForms2Blazor.Tests/FileConverterTests.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        {
+            FileConverter fc = new StaticFileConverter(_testStaticFilePath);
+
+            IEnumerable<FileInformation> fileList = await fc.MigrateFileAsync();
+            FileInformation fi = fileList.First();
+            byte[] bytes = fi.FileBytes;
+
+            Assert.IsTrue(bytes.Length != 0);
+            Assert.IsTrue(fi.RelativePath.Equals("wwwroot\\TestingArea\\TestFiles\\SampleStaticFile.png"));
+        }
```
Comment (jonlouie) [inline tst/CTA.WebForms2Blazor.Tests/FileConverterTests.cs:55]: A couple of suggestions for this test:

1. Rename test to `MigrateFileAsync_Moves_All_File_Bytes_To_wwwroot_Directory`. The ideal test name will include the function being tested, the condition of the test, and the expected outcome (e.g. `MyFunction_ThrowsException_With_Null_Parameter()`). Sometimes this is impossible or too wordy, so we can compromise based on context.
2. In the first assertion, we want to specify the exact number of bytes (e.g. what if the original file is 1 byte but the migrated file is 2 bytes?). Get the number of bytes of the original file and compare that to `fi.FileBytes.Length`.
3. Use `Path.Combine()` in the second `Assert` statement
4. Consider creating 1 test file per class, e.g. StaticFileConverterTests.cs. This will keep tests organized and prevent confusion if each converter class tests a shared interface with different implementations.

File: src/CTA.WebForms2Blazor/FileInformationModel/ConfigFileInformation.cs [hunk 5]
```diff
@@ -1,25 +1,14 @@
 ﻿using System;
 using System.Collections.Generic;
 using System.Text;
-using System.Threading.Tasks;
 
 namespace CTA.WebForms2Blazor.FileInformationModel
 {
-    public class ConfigFileInformation : FileInformation
+    class ConfigFileInformation : FileInformation
```
Comment (jonlouie) [inline src/CTA.WebForms2Blazor/FileInformationModel/ConfigFileInformation.cs:7]: Should this class be public?

File: src/CTA.WebForms2Blazor/FileInformationModel/CodeFileInformation.cs [hunk 6]
```diff
@@ -1,51 +1,14 @@
...
 using System.Collections.Generic;
-using System.IO;
-using System.Threading.Tasks;
-using CTA.WebForms2Blazor.Services;
+using System.Text;
 
 namespace CTA.WebForms2Blazor.FileInformationModel
 {
-    public class CodeFileInformation : FileInformation
+    class CodeFileInformation : FileInformation
```
Comment (jonlouie) [inline src/CTA.WebForms2Blazor/FileInformationModel/CodeFileInformation.cs:7]: Should this class be public?

File: src/CTA.WebForms2Blazor/FileConverters/FileConverter.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+        public string RelativePath { get { return _relativePath; } }
+
+        protected FileConverter(string relativePath)
+        {
+            _relativePath = relativePath;
+        }
+
+        public abstract Task<IEnumerable<FileInformation>> MigrateFileAsync();
+
+        protected abstract byte[] GetFileBytes();
```
Comment (jonlouie) [inline src/CTA.WebForms2Blazor/FileConverters/FileConverter.cs:20]: Have you and @comsemer discussed if getting the file bytes will differ based on the type of file being processed? My thought is that it would be identical for all files in which case `GetFileBytes()` can become a public method with an implementation. This would save work on each child class of `FileConverter`

File: src/CTA.WebForms2Blazor/Factories/FileConverterFactory.cs [hunk 8]
```diff
@@ -36,29 +36,29 @@ public FileInformation Build(FileInfo document)
             string relativePath = Path.GetRelativePath(_sourceProjectPath, document.FullName);
             string extension = document.Extension;
 
-            FileInformation fi;
+            FileConverter fi;
```
Comment (jonlouie) [inline src/CTA.WebForms2Blazor/Factories/FileConverterFactory.cs:39]: Rename `fi` to `fc` to match previous naming convention or to `fileConverter` as another suggestion.

**Review by jonlouie — CHANGES_REQUESTED** [review]

File: src/CTA.WebForms2Blazor/FileConverters/CodeFileConverter.cs [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+            _webFormsWorkspaceBuilder.NotifyNewExpectedDocument();
+
+            // TODO: Create class information objects and call
+            // _blazorProjectBuilder.NotifyNewExpectedDocument() in
+            // each one, this is because potentially multiple files
+            // are produced by this file information type
+        }
+
+        protected override byte[] GetFileBytes()
```
[the remark this answers is not in the document]
Comment (KallanBao) [reply src/CTA.WebForms2Blazor/FileConverters/CodeFileConverter.cs:32]: It won't let it be private since its inherited from an abstract class, the only way to make unaccessible from the other classes is making it protected

File: src/CTA.WebForms2Blazor/FileInformationModel/FileInformation.cs [hunk 10]
```diff
@@ -1,21 +1,22 @@
-﻿using System.Threading.Tasks;
+﻿using System;
 using System.Collections.Generic;
+using System.Text;
 
 namespace CTA.WebForms2Blazor.FileInformationModel
 {
     public abstract class FileInformation
```
[the remark this answers is not in the document]
Comment (KallanBao) [reply src/CTA.WebForms2Blazor/FileInformationModel/FileInformation.cs:7]: I mean different file information models can store different information right? Like config model might need to have separate information on which files its configurations are for

File: src/CTA.WebForms2Blazor/FileConverters/FileConverter.cs (hunk 7 above)
Comment (KallanBao) [reply src/CTA.WebForms2Blazor/FileConverters/FileConverter.cs:20]: I thought some of the other types of files we might do some processing before converting to byte[], but I can discuss again to confirm

File: src/CTA.WebForms2Blazor/FileInformationModel/FileInformation.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (comsemer) [reply src/CTA.WebForms2Blazor/FileInformationModel/FileInformation.cs:7]: Well the information they store is different, yes, but all of the information for the file is in bytes[], so the format of the information is the same. The point of the file information objects are to have a unified format to represent all files that the project builder can use indiscriminately

File: src/CTA.WebForms2Blazor/FileConverters/FileConverter.cs (hunk 7 above)
Comment (Eruanion) [reply src/CTA.WebForms2Blazor/FileConverters/FileConverter.cs:20]: We can also provide a base implementation and only override it if needed for exceptions. Another approach would be to store a base version of the implementation and reference that in all common method and use different code for the exceptions.

**Review by Eruanion — CHANGES_REQUESTED** [review]
Good work, just need to review other people's suggestions and comments.

File: src/CTA.WebForms2Blazor/MigrationManager.cs [hunk 11]
```diff
@@ -21,7 +21,7 @@ public class MigrationManager
 
         private WorkspaceManagerService _webFormsWorkspaceManager;
         private WorkspaceManagerService _blazorWorkspaceManager;
-        private FileInformationFactory _fileFactory;
+        private FileConverterFactory _fileFactory;
```
Comment (comsemer) [inline src/CTA.WebForms2Blazor/MigrationManager.cs:24]: I think Jon made a comment about changing the name to _converterFactory on the last version

**Review by comsemer — CHANGES_REQUESTED** [review]
Looks great, just that one field left that needs to be renamed

**Review by comsemer — APPROVED** [review]
Perfect

File: src/CTA.WebForms2Blazor/FileConverters/StaticFileConverter.cs [hunk 12]
```diff
@@ -15,18 +15,14 @@ public StaticFileConverter(string relativePath) : base(relativePath)
...
-        protected override byte[] GetFileBytes()
+        public override async Task<IEnumerable<FileInformation>> MigrateFileAsync()
         {
+            var newPath = Path.Combine("wwwroot", RelativePath);
             var workingDirectory = Environment.CurrentDirectory;
             var testProjectPath = Directory.GetParent(workingDirectory).Parent.Parent.FullName;
-            var path = Path.Combine(testProjectPath, RelativePath);
-            return File.ReadAllBytes(path);
-        }
+            var fullPath = Path.Combine(testProjectPath, RelativePath);
```
Comment (jonlouie) [inline src/CTA.WebForms2Blazor/FileConverters/StaticFileConverter.cs]: Does `testProjectPath` represent something from a test environment? If so, can we update this so our logic does not rely on a test environment?

**Review by jonlouie — COMMENTED** [review]
In addition to what @comsemer mentioned, I added one more question that I'm unsure of.

File: src/CTA.WebForms2Blazor/FileConverters/StaticFileConverter.cs (hunk 12 above)
Comment (KallanBao) [reply src/CTA.WebForms2Blazor/FileConverters/StaticFileConverter.cs]: Oh it doesn't rely on some test environment, it should just be the current directory. Its only named that way since I copy pasted the logic from somewhere else but I can fix that

File: src/CTA.WebForms2Blazor/FileConverters/StaticFileConverter.cs [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+        public StaticFileConverter(string relativePath) : base(relativePath)
+        {
+
+        }
+
+        public override async Task<IEnumerable<FileInformation>> MigrateFileAsync()
+        {
+            var newPath = Path.Combine("wwwroot", RelativePath);
+            var workingDirectory = Environment.CurrentDirectory;
+            var testProjectPath = Directory.GetParent(workingDirectory).Parent.Parent.FullName;
```
Comment (Eruanion) [inline src/CTA.WebForms2Blazor/FileConverters/StaticFileConverter.cs:22]: Also concerned about jon's comment here but this appears to be the context prefix for the path rather than actual test project path, perhaps just a naming oversight?

File: tst/CTA.WebForms2Blazor.Tests/Factories/FileConverterFactoryTests.cs [hunk 14]
```diff
@@ -5,13 +5,14 @@
...
+using CTA.WebForms2Blazor.FileConverters;
 
 namespace CTA.WebForms2Blazor.Tests.Factories
 {
     [TestFixture]
-    class FileInformationFactoryTests
+    class FileConverterFactoryTests
     {
-        private const string TestFilesDirectoryPath = "TestingArea/TestFiles";
+        private const string TestFilesDirectoryPath = "TestingArea\\TestFiles";
```
Comment (Eruanion) [inline tst/CTA.WebForms2Blazor.Tests/Factories/FileConverterFactoryTests.cs:15]: can we use path.combine as Jon suggested before so we can support multiple platforms?

**Review by Eruanion — CHANGES_REQUESTED** [review]
Just a couple of things left.

File: src/CTA.WebForms2Blazor/FileConverters/StaticFileConverter.cs (hunk 13 above)
Comment (KallanBao) [reply src/CTA.WebForms2Blazor/FileConverters/StaticFileConverter.cs:22]: Yep, I'll rename it!

File: src/CTA.WebForms2Blazor/FileConverters/StaticFileConverter.cs (hunk 12 above)
Comment (jonlouie) [reply src/CTA.WebForms2Blazor/FileConverters/StaticFileConverter.cs]: Got it, no problem. I think renaming it to reduce confusion should be sufficient then.

File: tst/CTA.WebForms2Blazor.Tests/Factories/FileConverterFactoryTests.cs (hunk 14 above)
Comment (KallanBao) [reply tst/CTA.WebForms2Blazor.Tests/Factories/FileConverterFactoryTests.cs:15]: Whoops, missed this one

**Review by Eruanion — APPROVED** [review]

**Review by jonlouie — APPROVED** [review]
